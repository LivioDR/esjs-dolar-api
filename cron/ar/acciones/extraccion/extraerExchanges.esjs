importar axios desde 'axios'
importar cheerio desde 'cheerio'
importar { parse } desde 'date-fns'
importar { collect } desde 'collect.js'

exportar asincrono funcion extraerExchanges() {
  const cotizaciones = esperar obtenerCotizaciones()

  retornar cotizaciones.mapear((cotizacion) => ({
    ...cotizacion,
    fechaActualizacion: crear Fecha(),
  }));
}

asincrono funcion obtenerCotizaciones() {
  retornar esperar Promesa.todos([
    consultarBelo(),
    consultarFiwind(),
    consultarCocos(),
    consultarDolarApp(),
    consultarTakenos(),
    consultarSatoshiTango(),
    consultarPrex(),
    consultarLemon(),
    consultarPlus(),
  ]).luego((respuestas) => {
    retornar respuestas.plano();
  });

  retornar cotizaciones
}

asincrono funcion consultarBelo() {
  const respuesta = esperar axios({
    url: 'https://api.belo.app/public/price',
  })

  retornar respuesta.data.mapear((cotizacion) => {
    const [parBase, parMoneda] = cotizacion.pairCode.dividir('/')

    const moneda = parMoneda.comienzaCon('USD') ? 'USD' : parMoneda
    const criptomoneda = parMoneda.comienzaCon('USD') ? parMoneda : nulo

    const monedaBase = parBase.comienzaCon('USD') ? 'USD' : parBase
    const criptomonedaBase = parBase.comienzaCon('USD') ? parBase : nulo

    retornar {
      exchange: 'belo',
      par: [monedaBase, moneda],
      parCriptomoneda: [criptomonedaBase, criptomoneda],
      compra: Numero(cotizacion.ask),
      venta: Numero(cotizacion.bid),
    };
  });
}

asincrono funcion consultarFiwind() {
  const respuesta = esperar axios({
    url: 'https://api.fiwind.io/v1.0/payments/pix/prices',
  })

  const brlToUsdt = respuesta.data.buscar(
    (cotizacion) => cotizacion.symbol === 'BRLUSDT',
  )

  const brlToArs = respuesta.data.buscar(
    (cotizacion) => cotizacion.symbol === 'BRLARS',
  )

  retornar [
    {
      exchange: 'fiwind',
      par: ['BRL', 'USD'],
      parCriptomoneda: ['USDT', nulo],
      compra: Numero(brlToUsdt.buy),
      venta: Numero(brlToUsdt.sell),
    },
    {
      exchange: 'fiwind',
      par: ['BRL', 'ARS'],
      compra: Numero(brlToArs.buy),
      venta: Numero(brlToArs.sell),
    },
  ];
}

asincrono funcion consultarCocos() {
  const respuesta = esperar (
    esperar fetch('https://api.cocos.capital/api/v1/public/pix-prices')
  ).json()

  retornar [
    {
      exchange: 'cocos',
      par: ['USD', 'BRL'],
      parCriptomoneda: ['USDT', nulo],
      compra: Numero(respuesta.usdToBrlRate),
      venta: nulo,
    },
    {
      exchange: 'cocos',
      par: ['BRL', 'ARS'],
      compra: Numero(respuesta.brlToArsRate),
      venta: nulo,
    },
  ];
}

asincrono funcion consultarDolarApp() {
  const respuesta = esperar axios({
    url: 'https://api.dolarapp.com/v1/tickers?currencies=ARS',
  })

  retornar respuesta.data.mapear((cotizacion) => ({
    exchange: 'dolarapp',
    par: ['USD', 'ARS'],
    parCriptomoneda: [cotizacion.book.dividir('_')[0].aMayusculas(), nulo],
    compra: Numero(cotizacion.ask),
    venta: Numero(cotizacion.bid),
  }));
}

asincrono funcion consultarTakenos() {
  const respuestaCompra = esperar axios({
    url: 'https://app.takenos.com/api/rates/BRL_USD?provider=kamiPay&operation=sell',
  })

  const respuestaVenta = esperar axios({
    url: 'https://app.takenos.com/api/rates/BRL_USD?provider=kamiPay&operation=buy',
  })

  retornar [
    {
      exchange: 'takenos',
      par: ['USD', 'BRL'],
      parCriptomoneda: [nulo, 'USDT'],
      compra: Numero(respuestaCompra.data.rate),
      venta: Numero(respuestaVenta.data.rate),
    },
  ];
}

asincrono funcion consultarSatoshiTango() {
  const respuesta = esperar axios({
    url: 'https://api.satoshitango.com/v3/qr/ticker',
  })

  retornar respuesta.data.data.tickers.mapear((cotizacion) => {
    const [parBase, parMoneda] = cotizacion.pair.dividir('/')

    const monedaBase = parBase.comienzaCon('USD') ? 'USD' : parBase
    const moneda = parMoneda.comienzaCon('USD') ? 'USD' : parMoneda

    retornar {
      exchange: 'satoshitango',
      par: [monedaBase, moneda],
      parCriptomoneda: [
        parBase.comienzaCon('USD') ? parBase : nulo,
        parMoneda.comienzaCon('USD') ? parMoneda : nulo,
      ],
      compra: Numero(cotizacion.rate),
      venta: nulo,
    };
  });
}

asincrono funcion consultarPrex() {
  const respuesta = esperar axios({
    url: 'https://api.prexcard.com.ar/v1/partners/pix-webs/quotes',
    headers: {
      'API-Key': import.meta.env.VITE_PREX_API_KEY,
    },
  })

  retornar respuesta.data.quotes.filtrar(
      (cotizacion) =>
        cotizacion.symbol === 'BRLUSD' || cotizacion.symbol === 'BRLARS',
    ).mapear((cotizacion) => {
      const [monedaBase, moneda] =
        cotizacion.symbol === 'BRLUSD' ? ['USD', 'BRL'] : ['BRL', 'ARS']

      retornar {
        exchange: 'prex',
        par: [monedaBase, moneda],
        compra: Numero(cotizacion.buy),
        venta: nulo,
      };
    });
}

asincrono funcion consultarLemon() {
  const respuesta = esperar axios({
    url: 'https://api.lemoncash.com.ar/api/v1/pix-payments/external-info',
  })

  retornar respuesta.data.pix_exchange_rates.mapear((cotizacion) => {
    const monedaBase = cotizacion.base_currency
    const moneda = cotizacion.quote_currency

    retornar {
      exchange: 'lemon',
      par: [monedaBase, moneda],
      compra: Numero(cotizacion.applied_price),
      venta: nulo,
    };
  });
}

asincrono funcion consultarPlus() {
  const respuesta = esperar axios({
    url: 'https://api.plus.com.ar/external/remittance/pix/cotization',
    headers: {
      Plusauthclient: 'comparapix',
      Plusauthkey: import.meta.env.VITE_PLUS_AUTH_KEY,
    },
  })

  retornar [
    {
      exchange: 'plus',
      par: ['USD', 'BRL'],
      parCriptomoneda: ['USDT', nulo],
      compra: Numero(respuesta.data.USDT_BRL),
      venta: nulo,
    },
    {
      exchange: 'plus',
      par: ['BRL', 'ARS'],
      compra: Numero(respuesta.data.ARS_BRL),
      venta: nulo,
    },
  ];
}
