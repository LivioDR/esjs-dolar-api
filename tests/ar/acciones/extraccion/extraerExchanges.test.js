import { describe, expect, it } from 'vitest'
import { extraerExchanges } from '@/ar/acciones/extraccion/extraerExchanges.esjs'

describe('extraerExchanges', () => {
  it('extrae cotizaciones de exchanges', async () => {
    const cotizaciones = await extraerExchanges()


    expect(cotizaciones.length).toBeGreaterThan(0)
    for (const cotizacion of cotizaciones) {
      expect(cotizacion).toBeTypeOf('object')

      if (cotizacion.compra)
        expect(cotizacion.compra).toBeTypeOf('number')

      expect(cotizacion.venta).toBeTypeOf('number')
      expect(cotizacion.par).toBeTypeOf('object')

      if (cotizacion.parCriptomoneda)
        expect(cotizacion.parCriptomoneda).toBeTypeOf('object')

      expect(cotizacion.fechaActualizacion).toBeTypeOf('object')
    }
  })
})
