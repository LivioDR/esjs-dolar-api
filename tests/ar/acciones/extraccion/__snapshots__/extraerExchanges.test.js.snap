// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extraerExchanges > extrae cotizaciones de exchanges 1`] = `
[
  {
    "exchange": "belo",
    "par": [
      "BTC",
      "ARS",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "ETH",
      "ARS",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "DAI",
      "ARS",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "USD",
      "ARS",
    ],
    "parCriptomoneda": [
      "USDT",
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "USD",
      "ARS",
    ],
    "parCriptomoneda": [
      "USDC",
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "BTC",
      "ETH",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "BTC",
      "DAI",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "BTC",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDT",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "BTC",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDC",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "ETH",
      "DAI",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "ETH",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDT",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "ETH",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDC",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "DAI",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDT",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "DAI",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDC",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "USD",
      "USD",
    ],
    "parCriptomoneda": [
      "USDT",
      "USDC",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "USD",
      "USD",
    ],
    "parCriptomoneda": [
      "USD",
      "USDT",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "BRL",
      "USD",
    ],
    "parCriptomoneda": [
      null,
      "USDT",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "BRL",
      "ARS",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "USD",
      "USD",
    ],
    "parCriptomoneda": [
      "USDT",
      "USD_AR",
    ],
  },
  {
    "exchange": "belo",
    "par": [
      "USD",
      "ARS",
    ],
    "parCriptomoneda": [
      "USD_AR",
      null,
    ],
  },
  {
    "exchange": "fiwind",
    "par": [
      "BRL",
      "USD",
    ],
    "parCriptomoneda": [
      "USDT",
      null,
    ],
  },
  {
    "exchange": "fiwind",
    "par": [
      "BRL",
      "ARS",
    ],
    "parCriptomoneda": undefined,
  },
  {
    "exchange": "cocos",
    "par": [
      "USD",
      "BRL",
    ],
    "parCriptomoneda": [
      "USDT",
      null,
    ],
  },
  {
    "exchange": "cocos",
    "par": [
      "BRL",
      "ARS",
    ],
    "parCriptomoneda": undefined,
  },
  {
    "exchange": "dolarapp",
    "par": [
      "USD",
      "ARS",
    ],
    "parCriptomoneda": [
      "USDC",
      null,
    ],
  },
  {
    "exchange": "takenos",
    "par": [
      "USD",
      "BRL",
    ],
    "parCriptomoneda": [
      null,
      "USDT",
    ],
  },
  {
    "exchange": "satoshitango",
    "par": [
      "USD",
      "ARS",
    ],
    "parCriptomoneda": [
      "USDT",
      null,
    ],
  },
  {
    "exchange": "satoshitango",
    "par": [
      "USD",
      "BRL",
    ],
    "parCriptomoneda": [
      "USDT",
      null,
    ],
  },
  {
    "exchange": "satoshitango",
    "par": [
      "BRL",
      "ARS",
    ],
    "parCriptomoneda": [
      null,
      null,
    ],
  },
]
`;
